src/condor_schedd.V6/schedd.cpp

```diff
@@ -6432,6 +6433,142 @@ UpdateGSICredContinuation::finish_update(ReliSock *rsock, ReliSock::x509_delegat
 }
 
 
+int
+Scheduler::getChildConctactInfo(int /*cmd*/, Stream* s) {
+	ClassAd request;
+	ReliSock* rsock = (ReliSock*)s;
+
+	// Set timeout to not block to long
+	rsock->timeout(10);
+	rsock->decode();
+
+	// make sure this connection is authenticated, and we know who
+	// the user is.
+	if ( ! rsock->triedAuthentication()) {
+		CondorError errstack;
+		if ( ! SecMan::authenticate_sock(rsock, READ, &errstack)) {
+			// we failed to authenticate, we should bail out now
+			// since we don't know what user is trying to perform
+			// this action.
+			dprintf(D_AUDIT | D_ERROR_ALSO, *rsock,
+			        "getContactInfo() failed to authenticate user aborting: %s\n",
+			        errstack.getFullText().c_str());
+			refuse(s);
+			return FALSE;
+		}
+	}
+
+	const OwnerInfo* sock_owner = EffectiveUserRec(rsock);
+	int cluster;
+	int proc;
+	int dt_type;
+
+	if ( ! getClassAd(rsock, request) || ! rsock->end_of_message()) {
+		dprintf(D_AUDIT | D_ERROR_ALSO, *rsock, "Can't read command request ad from tool\n");
+		refuse(s);
+		return FALSE;
+	}
+
+	if ( ! request.LookupInteger(ATTR_CLUSTER_ID, cluster)) {
+		dprintf(D_AUDIT | D_ERROR_ALSO, *rsock, "Invalid command request: Missing ClusterId\n");
+		refuse(s);
+		return FALSE;
+	}
+
+	if ( ! request.LookupInteger(ATTR_PROC_ID, proc)) {
+		dprintf(D_AUDIT | D_ERROR_ALSO, *rsock, "Invalid command request: Missing ProcId\n");
+		refuse(s);
+		return FALSE;
+	}
+
+	if ( ! request.LookupInteger("ContactDaemonType", dt_type)) {
+		dprintf(D_AUDIT | D_ERROR_ALSO, *rsock, "Invalid command request: Missing ProcId\n");
+		refuse(s);
+		return FALSE;
+	}
+
+	PROC_ID ID(cluster, proc);
+
+	ClassAd response;
+
+	JobQueueJob* job_ad = GetJobAd(ID);
+	shadow_rec* srec = FindSrecByProcID(ID);
+
+	bool success = false;
+	bool permitted = false;
+	int status;
+	int universe;
+	std::string addr;
+	std::string secret;
+	std::string reason;
+
+	if ( ! job_ad) {
+		reason = "Job not in queue";
+		goto send_response;
+	}
+
+	// Check that this user is allowed to talk to this DAGMan
+	if ( ! UserCheck2(job_ad, sock_owner)) {
+		reason = "Not permitted to talk to this job";
+		goto send_response;
+	}
+
+	permitted = true;
+
+	// Make sure this job is actually running (otherwise there is no daemon to contact)
+	if (GetAttributeInt(ID.cluster, ID.proc, ATTR_JOB_STATUS, &status) < 0 || status != RUNNING) {
+		reason = "Job is not currently running";
+		goto send_response;
+	}
+
+	switch (dt_type) {
+		case DT_DAGMAN:
+			// Contacting DAGMan... check this is a scheduler universe job and get contact addr
+			if (GetAttributeInt(ID.cluster, ID.proc, ATTR_JOB_UNIVERSE, &universe) < 0 || universe != CONDOR_UNIVERSE_SCHEDULER) {
+				reason = "Specified job is not scheduler universe";
+				goto send_response;
+			}
+
+			if (GetAttributeString(ID.cluster, ID.proc, ATTR_DAG_ADDRESS, addr) < 0 || addr.empty()) {
```

Comment:
The GetAttributeXXX() functions are useful only if you have a transaction open to modify the job queue or you don't have a pointer to the actual job ad. Here, you can do lookups directly on the job_ad, which is a little more efficient.